swan-lake/learn/writing-tests.md

```diff
@@ -0,0 +1,345 @@
+
+On the other hand, symbols defined in the test files will not be visible inside the module source files. When running tests, the symbols in the module source files will be initialized first followed by the ones in the test files.
+
+## Using Assertions
+
+The Ballerina test framework supports the following assertions which help verify the expected behaviour of a piece of code. These assertions can be used to decide if the test is passing or failing based on the condition.
+
+
+### **assertTrue(boolean expression, string message)**
+
+Asserts that the expression is true with an optional message.
+
+***Example:***
+```ballerina
+import ballerina/test;
+
+@test:Config {}
+function testAssertTrue() {
+    boolean value = false;
+    test:assertTrue(value, msg = "AssertTrue failed");
+}
+```
+
+### **assertFalse(boolean expression, string message)**
+
+Asserts that the expression is false with an optional message.
+
+***Example:***
+```ballerina
+import ballerina/test;
+
+@test:Config {}
+function testAssertFalse() {
+    boolean value = false;
+    test:assertFalse(value, msg = "AssertFalse failed");
+}
+```
+
+### **assertEquals(Any actual, Any expected, string message)**
+
+Asserts that the actual is equal to the expected, with an optional message.
+
+***Example:***
+```ballerina
+import ballerina/test;
+
+@test:Config {}
+function testAssertIntEquals() {
+
+    int answer = 0;
+    int a = 5;
+    int b = 3;
+    answer = intAdd(a, b);
+    test:assertEquals(answer, 8, msg = "IntAdd function failed");
+}
+
+function intAdd(int a, int b) returns (int) {
+    return (a + b);
+}
+```
+
+### **assertNotEquals(Any actual, Any expected, string message)**
+
+Asserts that the actual is not equal to the expected, with an optional message.
+
+***Example:***
+
+```ballerina
+import ballerina/test;
+
+@test:Config {}
+function testAssertIntEquals() {
+
+    int answer = 0;
+    int a = 5;
+    int b = 3;
+    answer = intAdd(a, b);
+    test:assertNotEquals(answer, 8, msg = "Matches");
+}
+
+function intAdd(int a, int b) returns (int) {
+    return (a + b);
+}
+```
+
+### **assertFail(string message)**
+
+Fails the test. Useful when we want to fail a test while in execution based on a check for a condition.
+
+***Example:***
+
+```ballerina
+import ballerina/test;
+
+@test:Config {}
+function foo() {
+    error? e = trap bar(); // Expecting `bar()` to panic
+    if (e is error) {
+        test:assertEquals(e.reason(), "Invalid Operation", msg = "Invalid error reason"); // Some other assertions
+    } else {
+        test:assertFail(msg = "Expected an error");
+    }
+}
+```
+
+## Setup and Teardown
+
+Following test annotations can be used for setup and teardown instructions. These annotations enable executing instructions in different levels.
+
+### @test:BeforeSuite {}
+
+The function specified following the annotation will be run once before any of the tests in the test suite is run. This can be used for initializing test suite level aspects.
```

Comment:
```suggestion
The function specified following the annotation will be run once before any of the tests in the test suite is run. This can be used for initializing the test suite level aspects.
```